cmd/snapd-generator/main.c

```diff
@@ -30,19 +31,192 @@
+		fprintf(stderr, "cannot create symlink %s: %m\n", lnpath);
+		return 1;
+	}
+
+	return 0;
+}
+
+#define MAJOR_LOOP_DEV 7
+#define SNAPD_DRIVERS_TREE_DIR "/var/lib/snapd/kernel"
+#define FIRMWARE_DIR "firmware"
+#define MODULES_DIR "modules"
+#define FIRMWARE_MNTPOINT "/usr/lib/" FIRMWARE_DIR
+#define MODULES_MNTPOINT "/usr/lib/" MODULES_DIR
+
+static int ensure_kernel_drivers_mounts(const char *normal_dir)
+{
+	const char* const kern_mnt_dir = "/run/mnt/kernel";
+	// Find mount information
+	sc_mountinfo *mounts SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
+	mounts = sc_parse_mountinfo("/proc/1/mountinfo");
+	if (!mounts) {
+		fprintf(stderr, "cannot open or parse /proc/1/mountinfo\n");
+		return 1;
+	}
+
+	// Create mount units only if not already present (which would be the
+	// case for an old initramfs) - otherwise systemd-fstab-generator
+	// complains, and older initramfs won't come in a kernel snap with
+	// support for components anyway.
+	const char* const kern_mntpts[] = { FIRMWARE_MNTPOINT, MODULES_MNTPOINT };
+	for (size_t i = 0; i < sizeof kern_mntpts/sizeof (char *); ++i) {
+		sc_mountinfo_entry *minfo = find_dir_mountinfo(mounts, kern_mntpts[i]);
+		// If the mounts already exist (old initramfs), do not create them -
+		// note that we additionally check for SNAPD_DRIVERS_TREE_DIR in the
+		// mount source to make sure the units created here are still
+		// generated on "systemctl daemon-reload".
+		if (minfo && strstr(minfo->root, SNAPD_DRIVERS_TREE_DIR) == NULL) {
+			return 0;
+		}
+	}
+
+	// Find active kernel name and revision by looking at what was
+	// mounted in /run/mnt/kernel by snap-bootstrap.
+
+	sc_mountinfo_entry *kern_minfo = find_dir_mountinfo(mounts, kern_mnt_dir);
+	if (!kern_minfo) {
+		// This is not Ubuntu Core / hybrid, do nothing and do not fail
+		return 0;
+	}
+
+	// Mount source should be a snap
+	if (!sc_streq(kern_minfo->fs_type, "squashfs")) {
+		fprintf(stderr, "unexpected fs type (%s) for %s\n",
+			kern_minfo->fs_type, kern_mnt_dir);
+		return 1;
+	}
+
+	// We expect a loop device as source
+	if (kern_minfo->dev_major != MAJOR_LOOP_DEV) {
+		fprintf(stderr, "mount source %s for %s is not a loop device\n",
+			kern_minfo->mount_source, kern_mnt_dir);
+		return 1;
+	}
+
+	// Find out backing file
+	char fname[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(fname, sizeof fname,
+			 "/sys/dev/block/%u:%u/loop/backing_file",
+			 kern_minfo->dev_major, kern_minfo->dev_minor);
+	FILE *f SC_CLEANUP(sc_cleanup_file) = NULL;
+	f = fopen(fname, "r");
+	if (!f) {
+		fprintf(stderr, "cannot open %s: %m\n", fname);
+		return 1;
+	}
+	char snap_path[PATH_MAX + 1] = { 0 };
+	if (fgets(snap_path, sizeof snap_path, f) == NULL) {
+		fprintf(stderr, "while reading %s: %m\n", fname);
+		return 1;
+	}
+
+	// Now parse the snap path
+	size_t i;
+	for (i = strlen(snap_path); i > 0 && snap_path[--i] != '/';);
+	char *snap_fname = snap_path + i + 1;
+
+	// snap_fname is of form <name>_<rev>.snap
```

Comment:
from the man of fgets snap_fname will/could also contain a newline?  it doesn't matter given that we ignore and don't check the extension, but I think it would be good if the comments/or the code accounted for that

Reply:
Good point, I've added a comment about this.